### PR TITLE
WIP: Database upgrades

### PR DESCRIPTION
**Pull request by ReCore-sys**
This PR provides many changes to the database system, primarily removing the multiple database options and using only LMDB through the [heed](https://docs.rs/heed/0.20.5/heed/index.html) crate. We've had great success with this database before and multiple databases has gotten very difficult to manage.

## Description
This PR removes the other database options and sets LMDB as the only database backend. We used LMDB before and it worked very well, but when rewriting we tried some other databases and had some initial success but eventually they all had too many drawbacks (redb being unreliable, surrealkv and sled using heaps of memory, rockdb requiring long compilation steps, etc.) so it was decided to only use LMDB since it offered very little drawbacks at the cost of having minimal documentation. So far this has lead to a significant reduction in memory use and no noticeable change in performance, though LMDB does use a memory mapping strategy that causes extremely high virtual memory usage. This doesn't have any real effect since the OS will reclaim it as soon as it needs it and all it really represents is the maximum size of the database, but it make our stats look concerning.

This PR also implements simple caching with the [moka](https://docs.rs/moka) crate. Currently nothing is being bottlenecked by IO speeds, but this will alleviate a lot of those issues before they arise. The size of the cache is configurable with in the config file, with higher limits allowing for less disk accesses at the cost of higher memory use. The cache drops entries based on a chunk's size in memory and how long it's been since the entry was accessed. The current default is 20MB with a TTL of 60 seconds, meaning chunks are dropped from the cache if they have been there the longest when the cache exceeds 20MB of memory use, or if a chunk hasn't been accessed for 60 seconds. Increasing the size will allow for more chunks to sit in cache at once, while increasing the TTL will mean chunks can stay in cache longer without being dropped, allowing less frequently accessed chunks to maintain fast access speeds.

Some minor API and documentation changes have been made to the anvil parser as well.

## Motivation and Context
The database has been a source of annoyance for a while now, requiring different implementations of features for each database. This should allow for a more unified approach.
Caching was implemented to reduce the performance hit that disk reads was having.

## How has this been tested?
No tests for now, but proper database tests will be included in future commits.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (restructuring code, without changing its behavior)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Clippy passes with no warnings.

### CODE REVIEW

**Review by 0xnim — CHANGES_REQUESTED** [review]
Just these changes. And that question

File: src/lib/adapters/anvil/src/lib.rs [hunk 1]
```diff
@@ -72,6 +72,10 @@ pub fn load_anvil_file(file_path: PathBuf) -> Result<LoadedAnvilFile, AnvilError
     let file = std::fs::File::open(&file_path)
         .map_err(|e| AnvilError::UnableToReadFile(file_path.clone(), e))?;
 
+    // Memory mapping is inherently unsafe since it's basically telling the os to use the file as,
```
[the remark this answers is not in the document]
Comment (ReCore-sys) [reply src/lib/adapters/anvil/src/lib.rs:75]: On windows at least I wasn't able tot delete the database files while the server was running, but I'm not if that applies to other OSs

**Comment by 0xnim** [comment]
Yeah its good

**Comment by vortexreanimated** [comment]
can i just fucking ask who we is??
> We've had great success with this database before and multiple databases has gotten very difficult to manage.

**Comment by ReCore-sys** [comment]
> can i just fucking ask who we is??
> 
> > We've had great success with this database before and multiple databases has gotten very difficult to manage.

First off, chill.
Secondly, we have tried a few databases in the past and LMDB is the only one that hasn't given us more headaches than it's worth. It's fast, uses minimal memory and doesn't have any weird caveats so it's a pretty good choice.

**Comment by Sweattypalms** [comment]
@vortexreanimated lmaooo "we" as in the people who worked for ferrumc, duh?

LMDB worked pretty good last time: it has low memory usage, and has good access/write speeds. 
Although if we see better options, we would most likely change our db then.

**Pull request merged**